Make key counter skinnable

ItsShamed opened the pull request:
https://github.com/ppy/osu/assets/39100084/c4510368-d59e-41ae-bf8f-01b164084666

A lot of things have changed in this PR, and I couldn't separate them into individual PRs without scraping some context right here. Hopefully, the diff will not be hard to read.

### Key counter logic refactor

This PR, refactors how the key counters work. The whole trigger management and count tracking have been decoupled from the actual key counters so that they can keep a consistent state regardless of when they are added mid-replay:
- The Receptor logic has been moved from `KeyCounterDisplay` to a new `KeyCounterController` class.
- The count logic has been moved from `KeyCounter` to `InputTrigger`.

### Attachment flow

The attachment flow has been abstracted out to support any component that would be attachable in the future (less boilerplate code).

### Test failures

One of the bottlenecks to making this PR are the existing tests that were relying too much on the old way of displaying the key counter. While most of them have been worked around, some tests still rely on old logic that is no longer present due to refactors in this PR and thus fail.

ItsShamed committed:
refactor: improve attachement flow

This removes the hard reliance on individual classes and makes its usage easiser to ingreate anywhere
else.

ItsShamed committed:
feat: integrate attachment flow in `SkinComponentsContainer`

ItsShamed committed:
test: remove hard usages of `KeyCounterDisplay`

ItsShamed committed:
refactor: move count logic in `InputTrigger`

This will allow us to keep track of the real count regardless of whether
the key counter has been placed mid-replay or not.

ItsShamed committed:
refactor: decouple Trigger logic from `KeyCounterDisplay`

This allows to keep a coeherent state regardless of the progress of the play

ItsShamed committed:
feat: make `KeyCounterDisplay` skinnable

ItsShamed committed:
revert: remove attachment logic from `SkinComponentContainer`

This is no longer in the scope of the PR.

ItsShamed committed:
Merge branch 'master' into hud/kc-skinnable

ItsShamed committed:
fix: change key counter position in Triangles and Legacy skins

ItsShamed committed:
test: move back key counter tests in ctor

ItsShamed committed:
refactor: remove detachment logic

No real use case, cleaning up the diff

ItsShamed committed:
style: remove unused IAttachableSkinComponent on KCD

ItsShamed committed:
style(KeyCounter): remove useless `Content` override

ItsShamed committed:
style(KeyCounter): remove useless `IsCounting` bindable

Counting logic has been moved to the `Trigger`

ItsShamed committed:
fix: correct key counter position in Triangles and Legacy skins

ItsShamed committed:
Merge branch 'master' into hud/kc-skinnable

ItsShamed committed:
test: change hideTarget drawable and testing logic

Doesn't change what it needs to test conceptually

ItsShamed commented:
I've made changes to the "failing" tests, hopefully I didn't change what they were supposed to test conceptually.

ItsShamed committed:
Merge branch 'master' into hud/kc-skinnable

ItsShamed committed:
test: add skin deserialisation test resource

ItsShamed committed:
test: adapt touch input test to changes

ItsShamed committed:
test: fix `KeyCounterController` not provided as a dependency

ItsShamed commented:
Failures on `TestSceneOsuTouchInput` seem relevant, even if in theory they should not. I can't figure out what broke here.

ItsShamed committed:
test(SkinDeserialisationTest): use more complete skin

ItsShamed committed:
test(TestSceneOsuTouchInput): simplify draw hierarchy

An InputTrigger is considered active as long as

ItsShamed committed:
test(TestSceneOsuTouchInput): fix `InputTrigger` depth

ItsShamed committed:
Merge branch 'master' into hud/kc-skinnable

bdach commented on osu.Game/Screens/Play/HUD/KeyCounterController.cs:
```diff
@@ -0,0 +1,95 @@
+
+        public override bool HandlePositionalInput => receptor == null;
+
+        public partial class Receptor : Drawable
```
If `KeyCounterController` is supposed to be a component that resides in the drawable hierarchy, is independent of skinnables, but is reused by said skinnables so that skins can be changed during gameplay without losing the key counts, then why is `Receptor` a thing at all still? Isn't it a useless layer of abstraction by now? Can't `KeyCounterController` be handling these things on its own now?

ItsShamed replied:
141f9efa

bdach commented on osu.Game/Screens/Play/HUD/KeyCounterDisplay.cs:
```diff
@@ -72,49 +55,31 @@ public void Add(InputTrigger trigger)
-                return base.Handle(e);
-            }
+            base.Dispose(isDisposing);
+            controller.OnNewTrigger -= Add;
```
When unsubscribing from an event handler on a BDL'd component, you need to `if (controller.IsNotNull())` guard the unsubscription, as there is a possibility that the drawable gets disposal called on it before it even loads (so it will never have dependencies injected into it, thus the nullref death).

ItsShamed replied:
1a8219ad

bdach reviewed (changes requested):
Before I even get into the nitty-gritty, serious doubts about the structure here as it's becoming overly complex and borderline inscrutable

ItsShamed committed:
style: guard event handler unsubscriptions

ItsShamed committed:
style(KeyCounterController): remove reliance on `Receptor`

ItsShamed committed:
refactor: tidy up attachement flow

TODO: find better naming and improve XMLDocs

ItsShamed commented on osu.Game/Rulesets/UI/RulesetInputManager.cs:
```diff
@@ -266,8 +283,11 @@ public interface IHasRecordingHandler
+
+    public interface IAttachableSkinComponent
+    {
     }
```
I tried to cook something up, however I'm not really inspired for the naming and the XMLDocs they essentially mean nothing right now.

bdach commented on osu.Game/Rulesets/UI/RulesetInputManager.cs:
```diff
@@ -158,69 +157,49 @@ protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
+            if (skinComponent.CanHandleKeybindings && skinComponent is Drawable component)
             {
-                foreach (var c
-                         in Target.Counters.Where(c => c.Trigger is KeyCounterActionTrigger<T>).Select(c => (KeyCounterActionTrigger<T>)c.Trigger))
-                    c.OnReleased(e.Action, Clock.Rate >= 0);
+                try
+                {
+                    KeyBindingContainer.Add(component);
+                    return;
+                }
+                catch (Exception)
+                {
+                    return;
+                }
```
This is again an overengineered and overfitted abstraction inferred from two separate usages, and there is no indication of any underlying structure here that would warrant defining any of this in this way. Loose sketch of how to fix this mess follows.

First of all, remove `IKeybindingEventsEmitter`. Replace it with:

```csharp
    public interface IHasKeyBindingContainer
    {
        KeyBindingContainer KeyBindingContainer { get; }
    }
```

Make `RulesetInputManager<T>` implement it. It'll just be one explicit interface implementation.

Secondly, remove _every member_ from `IKeyBindingListener` and replace with a single method:

```csharp
    public interface IKeyBindingAttachment
    {
        public void Attach(IHasKeyBindingContainer attachPoint);
    }
```

Thirdly, move _all_ setup logic of both attachable components to `IKeyBindingAttachment.Attach()`.

- In the case of the clicks per second calculator, `Attach()` should create the action listener on its own and add it to the keybinding container itself.
- In the case of the key counter, `Attach()` should read the keybindings from the container on its own and set up its children itself.

ItsShamed replied:
> In the case of the key counter, Attach() should read the keybindings from the container on its own and set up its children itself.

This would mean that the `KeyCounterController` would have to get access to the generic somehow (because of `KeyCounterActionTrigger<T>`). However, both making the class and the method generic seems like a no-go since it would be called in `HUDOverlay` which has no access to that generic. 
This again means that some logic has to be implemented in `RulesetInputManager<T>` — which is afaik the only class to have access to the generic in our context — but according to your solution it should not be the case, so I'm a bit stuck right here.

bdach replied:
I see. Well then, it seems that the only solution available without adding more generics and more indirection - both of which are generally bad - is to revert all of the `IAttachableSkinComponent` stuff and to keep the shape of `ICanAttachHUDPieces` as it was so it can accept the `KeyCounterController` and it can accept the `ClicksPerSecondCalculator` and _nothing else_.

bdach commented:
Some test failures in changed code here also warrant further inspection.

ItsShamed committed:
revert: bring back old attachment flow

As discussed, this would bring more problems that anything.

Refs: 4c39708, f83a4f4

ItsShamed committed:
Merge branch 'master' into hud/kc-skinnable

ItsShamed commented:
For the test failures, some are not related to this PR, and for the ones that seem to be related, they seem to fail only on CI:

![image](https://github.com/ppy/osu/assets/39100084/ea444058-2848-4df6-9f8f-078d74786320)

peppy committed:
Merge branch 'master' into hud/kc-skinnable

peppy committed:
Split padding out into constant to fix weird looking math

peppy committed:
Improve positioning and positioning code clarity for argon / triangles implementations

peppy committed:
Move argon key counter to right to match stable expectations

peppy committed:
Actuall add composite component to hierarchy

peppy committed:
Remove unused interface

peppy committed:
Split out interfaces from `RulesetInputManager` and improve xmldoc

peppy commented on osu.Game.Tests/Skins/SkinDeserialisationTest.cs:
```diff
@@ -51,6 +51,8 @@ public class SkinDeserialisationTest
             // Covers player avatar and flag.
             "Archives/modified-argon-20230305.osk",
+            // Covers key counters
+            "Archives/modified-argon-pro-20230618.osk"
```
Not sure why this is pro, but I guess it doesn't really matter.

peppy committed:
Add xmldoc for `KeyCounterController`

ItsShamed commented on osu.Game/Skinning/ArgonSkin.cs:
```diff
@@ -166,8 +168,20 @@ public ArgonSkin(SkinInfo skin, IStorageResourceProvider resources)
+                                            // Hard to find this at runtime, so taken from the most expanded state during replay.
+                                            const float song_progress_offset_height = 36 + padding;
+
+                                            keyCounter.Anchor = Anchor.BottomRight;
+                                            keyCounter.Origin = Anchor.BottomRight;
+                                            keyCounter.Position = new Vector2(-(hitError.Width + padding), -(padding * 2 + song_progress_offset_height));
```
Does it means that we are discarding the Figma design? It was expected that the key counter was positioned to the left and another Argon variant of the judgment counter to the right according to the design. 

![IMG_20230626_094407.jpg](https://github.com/ppy/osu/assets/39100084/3293c9c7-3433-4259-b4b1-53569a0a6693)

![IMG_20230626_094648.jpg](https://github.com/ppy/osu/assets/39100084/f4aacd0b-44cb-402d-82a1-2417ae584c26)

What is better between stable expectations and the Figma design feels blurry to me tbh.

peppy commented:
I'd propose either

`KeyCounterController` -> `InputCountController`
`JudgementTally` -> `JudgementCountController`

or

`KeyCounterController` -> `InputTally`
`JudgementTally`

or any other naming scheme where both of these classes use the same name.

peppy reviewed (changes requested):
Make quite a few changes. Seems okay but need to fix naming.

bdach committed:
Mark `IHasRecordingHandler.Recorder` nullable

bdach committed:
Fix attempting to add key counter controller to hierarchy in multiple places

bdach committed:
Remove redundant guard

In this particular case guarding with `.IsNull()` makes no sense, as the
`Trigger` is supplied in constructor and cannot feasibly be null. Doing
that only makes sense in scenarios where BDL / dependency injection is
involved.

bdach committed:
Group input handling members together

bdach committed:
Rename `{KeyCounter -> InputCount}Controller`

bdach committed:
Rename `Judgement{Tally -> CountController}`

bdach commented:
I've fixed code quality inspections, fixed tests, and applied renames as suggested in https://github.com/ppy/osu/pull/23926#issuecomment-1606902253.

@peppy please double check my changes. Of particular note is https://github.com/ppy/osu/pull/23926/commits/dbd76c1193bf545342e39cb67f997e6561224f5b, which could be considered a bit sketch - `HUDOverlay` creates and caches the controller, but attaches it in a completely different place in the hierarchy. (Before that change it was being added to _two_ places in the hierarchy, causing the test fails.) Initially I scoffed at this and tried moving the thing to `DrawableRuleset`, but (a) it felt bad putting it there layering-wise, and (b) it wasn't an easy change because some tests depend on it being in `HUDOverlay` specifically so I abandoned that course. Let me know if you think putting it in `DrawableRuleset` would be preferable after all.

bdach committed:
Fix incorrect indent size

bdach committed:
Attempt to remedy HUD overlay test failure by waiting more

peppy committed:
Add safety to tests to ensure loaded

peppy commented:
> Of particular note is [dbd76c1](https://github.com/ppy/osu/commit/dbd76c1193bf545342e39cb67f997e6561224f5b), which could be considered a bit sketch - `HUDOverlay` creates and caches the controller, but attaches it in a completely different place in the hierarchy.

~~While this might be fine with a comment, what gives with `clicksPerSecondCalculator`, which is seemingly still added in two places? I'm not sure how this doesn't outright `throw`.~~

Yeah okay this is not going to float. I'll try and figure something out as these two usages still are very different to each other and it smells very bad.

peppy committed:
Refactor `InputCountController` to not require being added to foreign body via `Attach`

I've made the flow match `ClicksPerSecondCalculator` as close as
possible. Hopefully this reads better.

peppy committed:
Move controllers above skinnable elements in initialisation order

peppy committed:
Change `JudgementCountController` to a `Component`

peppy committed:
Rename `ClicksPerSecondCalculator` to `ClicksPerSecondController`

peppy committed:
Change `JudgementCountController` to a `Component` and remove handling overrides

peppy commented:
@bdach back at you for a hopefully final pass.

bdach committed:
Attempt to fix more test failures

bdach committed:
Fix broken test step

bdach committed:
Merge branch 'master' into hud/kc-skinnable

bdach committed:
Add even more safety against unloaded components
